prediction_market_agent/db/blockchain_transaction_fetcher.py

```diff
@@ -0,0 +1,89 @@
+import polars as pl
+import spice
+from eth_typing import ChecksumAddress
+from prediction_market_agent_tooling.gtypes import xDai
+from prediction_market_agent_tooling.tools.hexbytes_custom import HexBytes
+from prediction_market_agent_tooling.tools.web3_utils import xdai_to_wei
+from web3 import Web3
+
+from prediction_market_agent.agents.microchain_agent.utils import decompress_message
+from prediction_market_agent.db.blockchain_message_table_handler import (
+    BlockchainMessageTableHandler,
+)
+from prediction_market_agent.db.models import BlockchainMessage
+from prediction_market_agent.utils import APIKeys
+
+
+class BlockchainTransactionFetcher:
+    MIN_TRANSACTION_AMOUNT = xDai(0.001)
+
+    def __init__(self) -> None:
+        self.blockchain_table_handler = BlockchainMessageTableHandler()
+
+    def unzip_message_else_do_nothing(self, data_field: str) -> str:
+        """We try decompressing the message, else we return the original data field."""
+        try:
+            return decompress_message(HexBytes(data_field))
+        except:
+            return data_field
+
+    def fetch_unseen_transactions_df(
+        self, consumer_address: ChecksumAddress
+    ) -> pl.DataFrame:
+        keys = APIKeys()
+        latest_blockchain_message = (
+            self.blockchain_table_handler.fetch_latest_blockchain_message(
+                consumer_address
+            )
+        )
+        min_block_number = (
+            0 if not latest_blockchain_message else latest_blockchain_message.block
+        )
+        # We order by block_time because it's used as partition on Dune.
+        # We use >= for block because we might have lost transactions from the same block.
+        # Additionally, processed tx_hashes are filtered out anyways.
+        query = f'select * from gnosis.transactions where "to" = {Web3.to_checksum_address(consumer_address)} AND block_number >= {min_block_number} and value > {xdai_to_wei(self.MIN_TRANSACTION_AMOUNT)} order by block_time asc'
```

Comment:
Any reason to be dependent on Dune? I asked o1 and it's possible to just use web3:

```python
from prediction_market_agent_tooling.tools.contract import ContractOnGnosisChain
from web3 import Web3
from prediction_market_agent_tooling.tools.web3_utils import xdai_to_wei
from prediction_market_agent_tooling.gtypes import ChecksumAddress, HexBytes, xDai


def get_transactions_to_address_trace(
    w3: Web3,
    target_address: ChecksumAddress,
    start_block: int,
    end_block: int | None = None,
    min_value: xDai | None = None,
) -> list[HexBytes]:
    if end_block is None:
        end_block = w3.eth.block_number

    trace_filter_params = {
        "fromBlock": hex(start_block),
        "toBlock": hex(end_block),
        "toAddress": [target_address],
    }
    min_value_wei = xdai_to_wei(min_value) if min_value is not None else None

    traces = w3.manager.request_blocking("trace_filter", [trace_filter_params])

    transactions = []
    for trace in traces:
        tx_hash = trace.get("transactionHash")
        if tx_hash:
            if min_value_wei is not None:
                action = trace.get("action")
                if action and "value" in action:
                    value = int(action["value"], 16)
                    if value < min_value_wei:
                        continue
                else:
                    # If value is missing in the trace, skip this transaction
                    continue
            transactions.append(HexBytes(tx_hash))

    return transactions


w3 = ContractOnGnosisChain.get_web3()

print(
    get_transactions_to_address_trace(
        w3, "0x184ca44A6c3cfc05bCF7246ac14101Ddb9423eAa", 37353329
    )
)
```

Reply:
Actually web3 can indeed fetch these transactions, but I was worried about RPC problems (which you also experienced during simulations).
I will try to use web3 instead of Dune (also my preferred option), also with a more popular address that has many transactions (e.g. market-creator). If all goes well, happy to use that instead of Dune.

Reply:
Unfortunately I don't think this will work - tried out with an address that has a lot more transactions (which is the use case we are building for) and found some READ timeouts from the RPC.
I don't like adding additional stuff as dependencies for no good reason, but in this case, adding Dune seems like the best choice.